utils/self-extr-exec/decompressor.cpp

```diff
@@ -0,0 +1,382 @@
+        /// Decompress data into file
+        if (0 != decompress(input, output, file_info.start, file_info.end))
+        {
+            if (0 != munmap(input, info_in.st_size))
+                perror(nullptr);
+            if (0 != munmap(output, file_info.uncompressed_size))
+                perror(nullptr);
+            return 1;
+        }
+
+        if (0 != fsync(output_fd))
+            perror(nullptr);
+        if (0 != close(output_fd))
+            perror(nullptr);
+    }
+
+    if (0 != munmap(input, info_in.st_size))
+        perror(nullptr);
+    return 0;
+}
+
+/// Copy particular part of command and update shift
+void fill(char * dest, char * source, size_t length, size_t& shift)
+{
+    memcpy(dest + shift, source, length);
+    shift += length;
+}
+
+/// Set command to `mv filename.decompressed filename && filename agrs...`
+void fillCommand(char command[], int argc, char * argv[], size_t length)
+{
+    memset(command, '\0', 3 + strlen(argv[0]) + 14 + strlen(argv[0]) + 4 + strlen(argv[0]) + length + argc);
+
+    /// position in command
+    size_t shift = 0;
+
+    /// Support variables to create command
+    char mv[] = "mv ";
+    char decompressed[] = ".decompressed ";
+    char add_command[] = " && ";
+    char space[] = " ";
+
+    fill(command, mv, 3, shift);
+    fill(command, argv[0], strlen(argv[0]), shift);
+    fill(command, decompressed, 14, shift);
+    fill(command, argv[0], strlen(argv[0]), shift);
+    fill(command, add_command, 4, shift);
+    fill(command, argv[0], strlen(argv[0]), shift);
+    fill(command, space, 1, shift);
+
+    /// forward all arguments
+    for (int i = 1; i < argc; ++i)
+    {
+        fill(command, argv[i], strlen(argv[i]), shift);
+        if (i != argc - 1)
+            fill(command, space, 1, shift);
+    }
+}
+
+int main(int argc, char* argv[])
+{
+    int input_fd = open(argv[0], O_RDONLY);
+    if (input_fd == -1)
+    {
+        perror(nullptr);
+        return 0;
+    }
+
+    bool have_compressed_analoge = false;
+
+    /// Decompress all files
+    if (0 != decompressFiles(input_fd, argv, have_compressed_analoge))
+    {
+        printf("Error happened");
+        if (0 != close(input_fd))
+            perror(nullptr);
+        return 1;
+    }
+
+    if (0 != close(input_fd))
+        perror(nullptr);
+
+    /// According to documentation `mv` will rename file if it
+    /// doesn't move to other directory.
+    /// Sometimes `rename` doesn't exist by default and
+    /// `rename.ul` is set instead. It will lead to errors
+    /// that can be easily avoided with help of `mv`
+
+    if (!have_compressed_analoge)
+    {
+        printf("Can't apply arguments to this binary");
+        /// remove file
+        char * name = strrchr(argv[0], '/') + 1;
+        execlp("rm", "rm", name, NULL);
+        perror(nullptr);
+        return 1;
+    }
+    else
+    {
+        /// move decompressed file instead of this binary and apply command
+        char bash[] = "/usr/bin/bash";
```

Comment:
But why cannot we run it directly, without bash?
ClickHouse should not require shell interpreter to run.

Reply:
We have to execute two programs: replace current binary with other file (`clickhouse.decompressed`) and execute it. I found solution only with bash. I'm open to ideas.

Reply:
Use `rename` function?

Reply:
From man:``` If newpath already exists, it will be atomically replaced, so that there is no point at which another process attempting to access newpath will find  it  missing.   However,  there will probably be a window in which both oldpath and newpath refer to the file being renamed.```
On practice it renames successfully, but `execve` tries to execute decompressor instead other file, and fails with `Text file busy`. But next command to this binary will be successful.